forc-pkg/src/pkg.rs

```diff
@@ -2090,12 +2090,92 @@ pub const SWAY_BIN_HASH_SUFFIX: &str = "-bin-hash";
 pub const SWAY_BIN_ROOT_SUFFIX: &str = "-bin-root";
 
 /// Builds a project with given BuildOptions
-pub fn build_with_options(build_options: BuildOptions) -> Result<Compiled> {
+pub fn build_with_options(build_options: BuildOptions) -> Result<()> {
+    let path = &build_options.path;
+
+    let this_dir = if let Some(ref path) = path {
+        PathBuf::from(path)
+    } else {
+        std::env::current_dir()?
+    };
+
+    // Decide the type of the ManifestFile at this_dir
+    if let Ok(package_manifest) = PackageManifestFile::from_dir(&this_dir) {
+        build_package_with_options(&package_manifest, build_options)?;
+    } else if let Ok(workspace_manifest) = WorkspaceManifestFile::from_dir(&this_dir) {
+        build_workspace_with_options(&workspace_manifest, build_options)?;
+    } else {
+        bail!(format!(
+            "Cannot find {} {:?}",
+            sway_utils::MANIFEST_FILE_NAME,
+            this_dir
+        ))
+    }
+    Ok(())
+}
+
+fn build_workspace_with_options(
+    manifest: &WorkspaceManifestFile,
+    build_options: BuildOptions,
+) -> Result<()> {
+    let workspace_dir = PathBuf::from(manifest.dir());
+    // Find root member of the workspace and build it.
+    let root_member = find_root_workspace_member(manifest, &build_options)?;
+    let root_member_path = workspace_dir.join(root_member);
+    let root_member_manifest = PackageManifestFile::from_dir(&root_member_path)?;
+    build_package_with_options(&root_member_manifest, build_options)?;
+    Ok(())
+}
+
+fn find_root_workspace_member(
+    manifest: &WorkspaceManifestFile,
+    build_options: &BuildOptions,
+) -> Result<String> {
+    let locked = build_options.locked;
+    let offline = build_options.offline_mode;
+    let members: HashSet<_> = manifest.members().cloned().collect();
+    for (member, member_path) in manifest.members().zip(manifest.member_paths()?) {
+        info!("Checking {member}");
+        let member_pkg_manifest = PackageManifestFile::from_dir(member_path.as_path())?;
+        let build_plan = BuildPlan::from_lock_and_manifest(&member_pkg_manifest, locked, offline)?;
+        let graph = build_plan.graph();
+        let discovered_deps = discover_contract_dependencies(graph, &members, member.clone())?;
+        // If we can find all members of the workspace from this member, this member is the root of the workspace.
+        // Meaning that building/deploying this member will ensure whole workspace will be handled in the
+        // correct order.
+        if discovered_deps.len() == members.len() {
```

Comment:
This approach works as long as there are no 2 sub-graphs that does not have any connection in between them. Meaning that any disjoint two sets present in the workspace may create problems. But if there are two disjoint sets of dependencies in the workspace, I am not sure if they should be in the same workspace. Because to have a situation like that there should be two sets of workspace members such that they do not have dependencies in between these two subsets including both `lib` and `contract` dependencies. So it comes to the relevance of having two totally independent group of projects in the same workspace. In my opinion that should be something that we warn the user about and not even allow but there might be some use cases that I am not aware of 🤔